### DISCUSSION TITLE
Persistent States and Memory Usage

### DISCUSSION THREAD

**Discussion by pantoast** [mrpmorris/Fluxor#374] [Q&A]
Hi,

I have been investigating using Fluxor for a fairly large application and I have been struggling to understand states and how they are persisted in memory. From what I have seen, all states are initialized as soon as the application is launched and are persisted through its lifetime. Would this not be a poor use of memory in a larger application? What if there were thousands or tens of thousands of states? In my case this is possible and it feels like it would be more efficient to have states scoped to the area they are relevant to, rather than all of them being global.

In looking for clearer understanding of this issue, I stumbled across this post from last year: https://github.com/mrpmorris/Fluxor/issues/215

I don't quite follow the answer though, particularly this:
> You can have separate chunks of state by adding multiple features. The idea is that each feature can be cleared down when you are in a different part of the app, in order to conserve memory.

What is meant by "each feature can be cleared down"? Is there a way to clear a feature and its state from memory? Or am I overreacting to this and having thousands of states in memory isn't actually a big deal?

Or, would IStateSelection with an immutable dictionary of states achieve what I'm looking for? Per this question: https://github.com/mrpmorris/Fluxor/issues/359

I could have a global state with an immutable Dictionary of sub-states that I can clear out of memory with effects, if that makes sense.

Thanks! Just looking for some clarification.

**Comment by pantoast** [comment]
Actually to follow up on my own question, I really should be thinking about Features not States. Features are being initialized immediately is more what I'm concerned about, not the States. The States are a side effect of the Features.

**Comment by mrpmorris** [comment]
When you agent in the X feature of your app, set its state to null (or some kind of empty state).

**Reply by pantoast** [reply]
I played around with it some more and it seems like having GetInitialState() return null works well. I then dispatch actions when components are Initialized and Disposed to initialize and dispose their related state. I think I am also blowing the issue out of proportion a little bit, if I had 50,000 states they still don't occupy much memory or take any significant time to initialize.

Thanks for the response!

**Reply by mrpmorris** [reply]
This is exactly why I named them "features" and not just state.

It's better to have the same data duplicated in different features' states and have reducers keep them in sync than it is to have a single copy of everything in a shared "state".

With features you know when the feature is no longer relevant in your user experience, and you can clear down its state to preserved memory without breaking something else showing the same data.

Without the feature approach you end up treating state like an RDBMS. One copy of each entity in memory that all your UX features read from, which means you can never know when it's safe to clear that state down. 

Also, if you hold local shared state for an entity that was deleted by another user on the server, you'll end up showing your user a deleted entity that you don't know has been deleted and can't safely remove from state if you do know.

Shared state is just one giant memory leak.
